rpc/src/v1/helpers/signing_queue.rs

```diff
@@ -14,86 +14,313 @@
+		// We don't really care about the result
+		let _ = self.sender.lock().unwrap().send(message);
+	}
+
+	fn remove(&self, id: U256) -> Option<TransactionConfirmation> {
+		self.queue.write().unwrap().remove(&id)
+	}
+
+	fn update_status(&self, id: U256, status: QueueStatus) {
+		let mut waiters = self.waiters.write().unwrap();
+		waiters.insert(id, status);
+	}
 }
 
-impl SigningQueue for ConfirmationsQueue {
+impl Drop for ConfirmationsQueue {
+	fn drop(&mut self) {
+		self.finish();
+	}
+}
+
+impl SigningQueue for  ConfirmationsQueue {
 	fn add_request(&self, transaction: TransactionRequest) -> U256 {
 		// Increment id
 		let id = {
 			let mut last_id = self.id.lock().unwrap();
 			*last_id = *last_id + U256::from(1);
 			*last_id
 		};
-		let mut queue = self.queue.lock().unwrap();
-		queue.insert(id, TransactionConfirmation {
-			id: id,
-			transaction: transaction,
-		});
+		// Add request to queue
+		{
+			let mut queue = self.queue.write().unwrap();
+			queue.insert(id, TransactionConfirmation {
+				id: id,
+				transaction: transaction,
+			});
+			debug!(target: "own_tx", "Signer: New transaction ({:?}) in confirmation queue.", id);
+		}
+		// Notify listeners
+		self.notify(QueueMessage::NewRequest(id));
 		id
 	}
 
-	fn remove_request(&self, id: U256) -> Option<TransactionConfirmation> {
-		self.queue.lock().unwrap().remove(&id)
+	fn peek(&self, id: &U256) -> Option<TransactionConfirmation> {
+		self.queue.read().unwrap().get(id).cloned()
+	}
+
+	fn request_rejected(&self, id: U256) -> Option<TransactionConfirmation> {
+		debug!(target: "own_tx", "Signer: Transaction rejected ({:?}).", id);
+		let o = self.remove(id);
+		self.update_status(id, QueueStatus::Rejected);
+		o
+	}
+
+	fn request_confirmed(&self, id: U256, hash: H256) -> Option<TransactionConfirmation> {
+		debug!(target: "own_tx", "Signer: Transaction confirmed ({:?}).", id);
+		let o = self.remove(id);
+		self.update_status(id, QueueStatus::Confirmed(hash));
+		o
 	}
 
 	fn requests(&self) -> Vec<TransactionConfirmation> {
-		let queue = self.queue.lock().unwrap();
+		let queue = self.queue.read().unwrap();
 		queue.values().cloned().collect()
 	}
+
+	fn wait_with_timeout(&self, id: U256) -> Option<H256> {
+		{
+			let mut waiters = self.waiters.write().unwrap();
+			let r = waiters.insert(id, QueueStatus::Waiting);
+			match r {
+				// This is ok, we can have many waiters
+				Some(QueueStatus::Waiting) | None => {},
+				// There already was a response for someone.
+				// The one waiting for it will cleanup, so...
+				Some(v) => {
+					// ... insert old status back
+					waiters.insert(id, v.clone());
+					if let QueueStatus::Confirmed(h) = v {
+						return Some(h);
+					}
+					return None;
+				},
+			}
+		}
+
+		info!(target: "own_tx", "Signer: Awaiting transaction confirmation... ({:?}).", id);
+		// Now wait for a response
+		let deadline = Instant::now() + Duration::from_secs(QUEUE_TIMEOUT_DURATION_SEC);
+		while Instant::now() < deadline {
```

Comment:
Not sure if it's worth the overhead of spinning up an extra thread.
Could you use: https://doc.rust-lang.org/stable/std/thread/fn.park_timeout.html and have whoever sets the status unpark the thread after it does so?